runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java

```diff
@@ -0,0 +1,230 @@
+import org.apache.beam.runners.core.construction.PipelineTranslation;
+import org.apache.beam.runners.core.construction.graph.ExecutableStage;
+import org.apache.beam.runners.core.construction.graph.FusedPipeline;
+import org.apache.beam.runners.core.construction.graph.GreedyPipelineFuser;
+import org.apache.beam.runners.fnexecution.GrpcFnServer;
+import org.apache.beam.runners.fnexecution.InProcessServerFactory;
+import org.apache.beam.runners.fnexecution.control.ProcessBundleDescriptors.SimpleProcessBundleDescriptor;
+import org.apache.beam.runners.fnexecution.control.SdkHarnessClient.ActiveBundle;
+import org.apache.beam.runners.fnexecution.control.SdkHarnessClient.BundleProcessor;
+import org.apache.beam.runners.fnexecution.control.SdkHarnessClient.RemoteInputDestination;
+import org.apache.beam.runners.fnexecution.control.SdkHarnessClient.RemoteOutputReceiver;
+import org.apache.beam.runners.fnexecution.data.GrpcDataService;
+import org.apache.beam.runners.fnexecution.logging.GrpcLoggingService;
+import org.apache.beam.runners.fnexecution.logging.Slf4jLogWriter;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.coders.BigEndianLongCoder;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.fn.channel.ManagedChannelFactory;
+import org.apache.beam.sdk.fn.stream.StreamObserverFactory;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.GroupByKey;
+import org.apache.beam.sdk.transforms.Impulse;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.WithKeys;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.values.KV;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Tests the execution of a pipeline from specification time to executing a single fused stage,
+ * going through pipeline fusion.
+ */
+@RunWith(JUnit4.class)
+public class RemoteExecutionTest implements Serializable {
+  private transient GrpcFnServer<FnApiControlClientPoolService> controlServer;
+  private transient GrpcFnServer<GrpcDataService> dataServer;
+  private transient GrpcFnServer<GrpcLoggingService> loggingServer;
+  private transient SdkHarnessClient controlClient;
+
+  private transient ExecutorService serverExecutor;
+  private transient ExecutorService sdkHarnessExecutor;
+
+  @Before
+  public void setup() throws Exception {
+    // Setup execution-time servers
+    ThreadFactory threadFactory = new ThreadFactoryBuilder().setDaemon(true).build();
+    serverExecutor = Executors.newCachedThreadPool(threadFactory);
+    InProcessServerFactory serverFactory = InProcessServerFactory.create();
```

Comment:
It would be valuable for test reasons to make a JUnit rule which spins up an SDK harness and tears it down as its useful in many tests. Separate out as a different PR.

Reply:
https://github.com/apache/beam/pull/4825